### ISSUE TITLE
set a client role without "Composite Roles"

### ISSUE THREAD

**Issue by Trois-Six** [mrparkers/terraform-provider-keycloak#154]
Hi,

I am using this version :

```$ dpkg -l | grep keycloak
ii  terraform-provider-keycloak    1.10.0+git20190907.f15366f     amd64    Terraform keycloak provider
```
When I try to set an openid connect client role without any composite role (an empty list), I have got an error:
```
resource "keycloak_role" "role_yyyy" {
  realm_id        = "${module.realm_xxxx.id}"
  client_id       = "${module.openid_client_xxxx_yyyy.id}"
  name            = "user"
  composite_roles = []
}
```
Plan:
```
Error: keycloak_role.role_yyyy: composite_roles: attribute supports 1 item as a minimum, config has 0 declared
```
I would like to be able to set an empty list to be sure of the conformity of my role.

Thank you,

Pierre

**Comment by mrparkers** [comment]
Hi @Trois-Six, the role will not have any composite roles if you omit the `composite_roles` attribute completely.  Let me know if that works for you.

**Comment by Trois-Six** [comment]
Yes that works, but if I have a role that have composite roles and I want to remove the composite roles, I can't.

**Comment by mrparkers** [comment]
Omitting the attribute _should_ remove all of that role's composites, can you give me the HCL you're using and explain the behavior you're getting verses the behavior you're expecting?

**Comment by mrparkers** [comment]
Also, my [acceptance tests explicitly cover this scenario](https://github.com/mrparkers/terraform-provider-keycloak/blob/master/provider/resource_keycloak_role_test.go#L236-L240) (adding composite roles and removing them all) so if you're noticing some strange behavior, it's probably a bug that I'd like to get fixed.

**Comment by Trois-Six** [comment]
OK I confirm that omitting the attribute remove all the role's composites. That's not usual for a Terraform resource but it works. Thank you !

**Comment by mrparkers** [comment]
I'm open to allowing users to specify an empty list of `composite_roles` if that would make you feel better.  I would also like to keep the behavior of removing composite roles if the attribute is not specified.

If this is something you want, you can re-open this issue and I'll work on it when I get a chance.

**Comment by Trois-Six** [comment]
The way it is done, I can't use this resource inside a module with a default value, but I can workaround it for the moment.

**Comment by mrparkers** [comment]
That's a fair use case, I did not think about that.

I think with HCL2 you could do something like this:

```hcl
resource "keycloak_role" "role" {
    realm_id = "master"
    name     = "test"

    composite_roles = len(var.composites == 0) ? null : var.composites
}
```

I think with HCL1 (which is what I'm still using) you'd have to do something even worse due to the lack of the `null` type:

```hcl
resource "keycloak_role" "role" {
    count = "${len(var.composites) == 0 ? 1 : 0}"
    realm_id = "master"
    name     = "test"
}

resource "keycloak_role" "role_with_composites" {
    count = "${len(var.composites) == 0 ? 0 : 1}"
    realm_id = "master"
    name     = "test"

    composite_roles = "${var.composites}"
}
```

In either case, I agree that it is silly to have to add logic to your HCL because the provider won't allow you to specify an empty list for this attribute.

Thanks for helping me understand your use case, I'll work on this when I have some free time.
